tests/pass/float.rs

```diff
@@ -148,155 +212,266 @@ fn basic() {
+            assert_eq_msg::<$fty>(
+                i as $fty,
+                f,
+                format_args!("{} -> {}: {msg}", stringify!($ity), stringify!($fty)),
+            );
+        }
+
+        /// Check both float to int and int to float
+        fn assert_bidir(f: $fty, i: $ity, msg: &str) {
+            assert_ftoi(f, i, msg);
+            assert_itof(i, f, msg);
+        }
+
+        /// Check both float to int and int to float
+        fn assert_bidir_unrep(f: $fty, i: $ity, msg: &str) {
+            assert_ftoi_unrep(f, i, msg);
+            assert_itof(i, f, msg);
+        }
+
+        let fbits = <$fty>::BITS;
+        let ibits = <$ity>::BITS;
+        let imax: $ity = <$ity>::MAX;
+        let imin: $ity = <$ity>::MIN;
+        let izero: $ity = 0;
+        #[allow(unused_comparisons)]
+        let isigned = (0 as $ity).saturating_sub(1) < 0;
+
+        #[allow(overflowing_literals)]
+        let imin_f: $fty = $imin_f;
+        #[allow(overflowing_literals)]
+        let imax_f: $fty = $imax_f;
+
+        // If an integer can fit entirely in the mantissa (counting the hidden bit), every value
+        // can be represented exactly.
+        let all_ints_exact_rep = ibits <= <$fty>::SIGNIFICAND_BITS + 1;
+
+        // Skip unchecked cast if negative numbers are unrepresentable
+        let assert_ftoi_neg = if isigned { assert_ftoi } else { assert_ftoi_unrep };
+        // Skip unchecked cast when int min/max would be unrepresentable
+        let assert_ftoi_big = if all_ints_exact_rep { assert_ftoi } else { assert_ftoi_unrep };
+        let assert_bidir_big = if all_ints_exact_rep { assert_bidir } else { assert_bidir_unrep };
+
+        // Near zero representations
+        assert_bidir(0.0, 0, "zero");
+        assert_ftoi(-0.0, 0, "negative zero");
+        assert_ftoi_neg(-0.99999999999999999999999999, izero.saturating_sub(1), "near -1");
+        assert_ftoi(<$fty>::from_bits(0x1), 0, "min subnormal");
+        assert_ftoi(<$fty>::from_bits(0x1 | 1 << (fbits - 1)), 0, "min neg subnormal");
+
+        // Spot checks. Use `saturating_sub` to create negative integers so that unsigned
+        // integers stay at zero.
+        assert_ftoi(0.9, 0, "0.9");
+        assert_ftoi_neg(-0.9, 0, "-0.9");
+        assert_ftoi(1.1, 1, "1.1");
+        assert_ftoi_neg(-1.1, izero.saturating_sub(1), "-1.1");
+        assert_ftoi(1.9, 1, "1.9");
+        assert_ftoi_neg(-1.9, izero.saturating_sub(1), "-1.9");
+        assert_ftoi(5.0, 5, "5.0");
+        assert_ftoi_neg(-5.0, izero.saturating_sub(5), "-5.0");
+        assert_ftoi(5.9, 5, "5.0");
+        assert_ftoi_neg(-5.9, izero.saturating_sub(5), "-5.0");
+
+        // Exercise the middle of the integer's bit range. A single digit fits as long as the
+        // exponent can fit its offset, so cap at the largest shift representable by the
+        // exponent.
+        let half_i_max: $ity = 1 << (ibits / 2).min((1 << (<$fty>::EXPONENT_BITS - 1)) - 1);
```

Comment:
I am quite confused by this... this is `1 << min(ibits / 2, (1 << (<$fty>::EXPONENT_BITS - 1)) - 1)`, right? The LHS is still the log but the RHS is already the power...?

I would have expected something more like `1 << min(ibits / 2, <$fty>::EXPONENT_BITS - 1)`.

Reply:
`1 << (EXPONENT_BITS - 1)` would be `1 << 4`, which is the max value that can fit in an exponent but not the max value the exponent can represent. However, turns out I was just recalculating `EXPONENT_BIAS`, so I can use that constant instead.

Reply:
Ah, yes makes sense, thanks. I forgot that the exponent is one more log level away, of course it is.